### PR TITLE
Add useful warnings for t.grad when it won't be populated for known reasons

### LINKED ISSUE #2362: raise a warning if someone tries to access .grad of a Variable that will never be computed
@soumith 
In my understanding, these two kinds of initializations should give the same result.

```
grid_data_variable = Variable(torch.rand(batch_size, gh, gw, d, nchans), requires_grad=True).cuda()
guide_data_variable = Variable(torch.rand(batch_size, h, w), requires_grad=True).cuda()
```

```
grid_data_variable = Variable(torch.rand(batch_size, gh, gw, d, nchans).cuda(), requires_grad=True)
guide_data_variable = Variable(torch.rand(batch_size, h, w).cuda(), requires_grad=True)
```

Then I test them on my customized layer.
The first one would fail in the **gradcheck** while the second works well.

Do you have any clues why the first one gives an incorrect gradient?

cc @ezyang @SsnL @albanD @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by albanD**
Fix #2362 and #19778

To avoid issues with frozen model, we only consider warning for Tensors that require gradients and are neither leafs nor retain gradients.

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 1a30902:
* **1/1** broken upstream at merge base 106ab48 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=10&sha1=106ab487ebbd1d7da9c3627e81b84b8f761ff8e8&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see [age history](https://dr.pytorch.org/master-viable-commits.html)):
      ```
      git fetch viable/strict
      git rebase viable/strict
      ```
* **0/1** failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=1a3090276f074cba7591e5e40c2fb8293e191f92):

### 1 failures recognized by patterns:
####  pytorch_linux_backward_compatibility_check_test
**Step:** Test
```
Nov 27 20:33:03 caused by: Connection refused (os error 111)
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: torch/tensor.py [hunk 1]
```diff
@@ -722,4 +722,13 @@ def _update_names(self, names, inplace):
         else:
             return super(Tensor, self).rename(names)
 
+    @property
+    def grad(self):
+        if not self.requires_grad:
```
Comment (gchanan) [inline torch/tensor.py]: won't these not show in C++/JIT?  I guess that's okay for now, but maybe file an issue?

File: torch/tensor.py [hunk 2]
```diff
@@ -722,4 +722,13 @@ def _update_names(self, names, inplace):
...
+            warnings.warn("Accessing .grad on a Tensor that does not require gradients always returns None.")
+        if not hasattr(self, "retains_grad") and not self.is_leaf:
```
Comment (gchanan) [inline torch/tensor.py]: wait, do both this and retain_grad above mean that I can't actually turn retain_grad off?

**Review by gchanan — APPROVED** [review]

File: torch/tensor.py (hunk 2 above)
Comment (albanD) [reply torch/tensor.py]: Given the implementation above, the hook cannot be removed.
So no we cannot remove the retain_grad. But that's fine as this is only used on temporary Tensors during the forward.

File: torch/tensor.py (hunk 1 above)
Comment (albanD) [reply torch/tensor.py]: I choose to put it here as otherwise it wouldn't know if this Tensor retains grad. Which is needed for an accurate message.

**Comment by vincentqb** [comment]
Discussed offline with @albanD: in the current form, warnings will be raised by the optimizers. This PR should update the optimizers so as to avoid such warnings. We can update the behaviour of optimizers in a subsequent PR.

File: test/test_jit.py [hunk 3]
```diff
@@ -13455,8 +13455,11 @@ def checkTracerWarning(self, *args, **kwargs):
         with warnings.catch_warnings(record=True) as warns:
             torch.jit.trace(*args, **kwargs)
         self.assertGreater(len(warns), 0)
+        found_warning = False
         for warn in warns:
-            self.assertIn("cause the trace to be incorrect", str(warn.message))
+            if "cause the trace to be incorrect" in str(warn.message):
```
Comment (gchanan) [inline test/test_jit.py]: can't you combine all of this into one line?

File: torch/tensor.py [hunk 4]
```diff
@@ -288,10 +288,10 @@ def trim(str):
 
     def retain_grad(self):
         r"""Enables .grad attribute for non-leaf Tensors."""
-        if self.grad_fn is None:  # no-op for leaves
-            return
         if not self.requires_grad:
             raise RuntimeError("can't retain_grad on Tensor that has requires_grad=False")
+        if self.is_leaf:  # no-op for leaves
```
Comment (gchanan) [inline torch/tensor.py]: probably worth noting that this is BC-breaking, although this would be a weird thing to do.  But just add "label: bc-breaking" and show a simple example.

File: torch/tensor.py [hunk 5]
```diff
@@ -722,4 +722,25 @@ def _update_names(self, names, inplace):
...
 
+    @property
+    def grad(self):
+        """
+        This attribute is ``None`` by default and becomes a Tensor the first time a call to
+        :func:`backward` computes gradients for ``self``.
+        The attribute will then contain the gradients computed and future calls to
+        :func:`backward` will accumulate (add) gradients into it.
+        """
+        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf:
```
Comment (gchanan) [inline torch/tensor.py]: I don't really understand how setting manually works.  I can always set it manually but if I try to access it when I haven't set requires_grad will warn me?

**Review by gchanan — APPROVED** [review]

File: torch/tensor.py (hunk 4 above)
Comment (albanD) [reply torch/tensor.py]: The change to `gradcheck` is a good example of the kind of BC change.
Basically calling `.retain_grad()` on something that is going to be ignored later.

File: torch/tensor.py (hunk 5 above)
Comment (albanD) [reply torch/tensor.py]: I reduced a lot the scope of the warning.
I will add a `and self._grad is None`.
So that it will only warn when you have a Tensor that requires gradient, is inside the graph, does not retain_grad and has no gradient set.

File: test/test_jit.py (hunk 3 above)
Comment (albanD) [reply test/test_jit.py]: Sure

File: test/test_torch.py [hunk 6]
```diff
@@ -5863,6 +5863,25 @@ def test_batch_norm_cpu_inference(self):
...
+
+            # Accessing .grad on non-leaf
+            dummy = dummy.clone()
+            foo = dummy.grad
+            self.assertEqual(len(w), 1)
+
+            # Accessing .grad on non-leaf that retain gradients
+            dummy.retain_grad()
+            foo = dummy.grad
+            self.assertEqual(len(w), 1)
```
Comment (vincentqb) [inline test/test_torch.py]: nit: Since you are just catching any warnings, there could be a situtation where an unrelated warning is raised but not the relevant one?

File: test/test_torch.py [hunk 7]
```diff
@@ -5863,6 +5863,25 @@ def test_batch_norm_cpu_inference(self):
...
+            dummy.requires_grad_()
+            foo = dummy.grad
+            self.assertEqual(len(w), 0)
+
+            # Accessing .grad on non-leaf
+            dummy = dummy.clone()
+            foo = dummy.grad
+            self.assertEqual(len(w), 1)
+
+            # Accessing .grad on non-leaf that retain gradients
```
Comment (vincentqb) [inline test/test_torch.py]: nit: retains?

File: torch/tensor.py [hunk 8]
```diff
@@ -722,4 +722,25 @@ def _update_names(self, names, inplace):
...
+    def grad(self):
+        """
+        This attribute is ``None`` by default and becomes a Tensor the first time a call to
+        :func:`backward` computes gradients for ``self``.
+        The attribute will then contain the gradients computed and future calls to
+        :func:`backward` will accumulate (add) gradients into it.
+        """
+        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf and self._grad is None:
+            warnings.warn("Accessing .grad on a Tensor that is not a leaf Tensor: its .grad attribute won't "
+                          "be populated during autograd.backward(). See .retain_grad() if you want the gradient to be saved.")
```
Comment (vincentqb) [inline torch/tensor.py]: I've googled "pytorch retain_grad" and didn't find more details about his. Maybe you could simply add the details to the pull request, and link to the pull request instead?

Without this, I suspect a user may not know how to fix this if there's something to fix after reading just "See .retain_grad() if you want the gradient to be saved."

Thoughts?

**Review by vincentqb — APPROVED** [review]
LGTM, but I wonder if the warning is clear enough in terms of
* what to do to fix if there is something to fix?
* what to do to suppress the warning if there isn't something to fix?

File: test/test_torch.py (hunk 6 above)
Comment (albanD) [reply test/test_torch.py]: Since we only call `retain_grad()` and `.grad` in valid cases. We actually don't want other warnings to be raised here. Otherwise, we have an overly verbose warning.

File: torch/tensor.py (hunk 8 above)
Comment (albanD) [reply torch/tensor.py]: The [doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.retain_grad) is indeed pretty limited.
Should I add an example of what it does to the doc and state explicitly to check the pytorch doc in the warning?

Comment (vincentqb) [reply torch/tensor.py]: I see you could add the information to anyone of the following, with an explicit link in the warning.

* #30531
* [doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.retain_grad)
* updated doc in #30253

The first seems natural since it links to other related issues and gives you space to write details. Thoughts?

Comment (albanD) [reply torch/tensor.py]: I updated the comment to link this PR.

**Comment by albanD** [comment]
> what to do to fix if there is something to fix?

Use `.retain_grad()` if you want the gradient for a non-leaf Tensor. Or make sure you have the leaf Tensor if your have a non-leaf Tensor by mistake.

> what to do to suppress the warning if there isn't something to fix?

I would argue there is always something to fix. If you always get this warning and want to suppress it, just replace the function call by `None` 😄

**Comment by vincentqb** [comment]
> > what to do to fix if there is something to fix?
> 
> Use `.retain_grad()` if you want the gradient for a non-leaf Tensor. Or make sure you have the leaf Tensor if your have a non-leaf Tensor by mistake.
> 
> > what to do to suppress the warning if there isn't something to fix?
> 
> I would argue there is always something to fix. If you always get this warning and want to suppress it, just replace the function call by `None` 😄

How about this?

The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). Please either use .retain_grad() if you want the gradient for a non-leaf Tensor, or make sure you have the leaf Tensor if you have a non-leaf Tensor by mistake. See PR #30531 for more informations.

File: torch/tensor.py [hunk 9]
```diff
@@ -728,4 +728,28 @@ def _update_names(self, names, inplace):
...
+        """
+        This attribute is ``None`` by default and becomes a Tensor the first time a call to
+        :func:`backward` computes gradients for ``self``.
+        The attribute will then contain the gradients computed and future calls to
+        :func:`backward` will accumulate (add) gradients into it.
+        """
+        if self.requires_grad and not hasattr(self, "retains_grad") and not self.is_leaf and self._grad is None:
+            warnings.warn("The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad "
+                          "attribute won't be populated during autograd.backward(). If you indeed want the gradient "
+                          "for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the "
```
Comment (gchanan) [inline torch/tensor.py]: just some nits (CC @vincentqb):

- "if you access" -> "if you accessed"

- "the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead"

There's not a general "the leaf Tensor" to access.  This sentence isn't really actionable, you would have to know _why_ they accessed to the non-leaf tensor to know, so I'd just get rid of it.

- We should link to documentation instead of pull requests.

**Comment by hadypranoto** [comment]
> > > what to do to fix if there is something to fix?
> > 
> > 
> > Use `.retain_grad()` if you want the gradient for a non-leaf Tensor. Or make sure you have the leaf Tensor if your have a non-leaf Tensor by mistake.
> > > what to do to suppress the warning if there isn't something to fix?
> > 
> > 
> > I would argue there is always something to fix. If you always get this warning and want to suppress it, just replace the function call by `None` 😄
> 
> Then the person has something to fix :) How about this?
> 
> The `.grad` attribute of a Tensor that is not a leaf Tensor is being accessed. Its `.grad` attribute won't be populated during `autograd.backward()`. If you indeed want the gradient for a non-leaf Tensor, use `.retain_grad()` on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See PR #30531 for more informations.

yes, sir vincent.

I'm also interested in that sentence too, if it has to be fixed, how do you fix it ... is there a sample code?
